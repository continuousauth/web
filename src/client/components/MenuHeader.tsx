import * as React from 'react';

import { Avatar, WarningSignIcon, Pane } from 'evergreen-ui';
import { UserState } from '../state/user';

import styles from './MenuHeader.scss';
import { CFALogo } from './icons/Logo';
import { cx } from '../utils';
import { Link, withRouter } from 'react-router-dom';

export const isPathActive = (path: string, exact = true) => {
  if (exact) return location.pathname === path;
  return location.pathname.indexOf(path) === 0;
};

export function MenuHeaderInner() {
  const user = React.useContext(UserState);

  return (
    <Pane className={styles.container}>
      <CFALogo className={styles.logo} />
      <span className={styles.productName}>CFA</span>
      <Link className={cx(styles.item, (isPathActive('/') && styles.active) || null)} to="/">
        Dashboard
      </Link>
      <a className={styles.item} href="https://docs.continuousauth.dev" target="_blank">
        Documentation
      </a>
      <div className={cx(styles.item, styles.warning)}>
        <WarningSignIcon color="warning" marginRight={8} />
        CFA is open-source and does not have formal support
      </div>
      <Pane className={styles.right}>
        <Avatar
          name={user ? user.displayName : '?'}
          size={24}
          isSolid
          src={user ? `https://github.com/${user.username}.png` : undefined}
        />
        <span>{user ? user.username : '?'}</span>
      </Pane>
      <span></span>
    </Pane>
  );
}

export const MenuHeader = withRouter(MenuHeaderInner);
