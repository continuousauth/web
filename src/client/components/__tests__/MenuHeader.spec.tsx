/**
 * @jest-environment jsdom
 */

import * as React from 'react';
import { shallow, mount } from 'enzyme';

import { MenuHeaderInner, isPathActive } from '../MenuHeader';
import { UserState } from '../../state/user';
import { BrowserRouter } from 'react-router-dom';

const fakeLocation = (path: string) => {
  delete (window as any).location;
  window.location = {
    pathname: path,
  } as any;
};

describe('<MenuHeader />', () => {
  it('Should render a logo', () => {
    const wrapper = shallow(<MenuHeaderInner />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('CFALogo')).toHaveLength(1);
  });

  it('should handle the user not being signed in', () => {
    const wrapper = shallow(<MenuHeaderInner />);
    const avatar = wrapper.find('Memo(ForwardRef(Avatar))');
    expect(avatar).toHaveLength(1);
    expect(avatar.prop('name')).toBe('?');
  });

  it('should show the users name when they are signed in', () => {
    function WithUser() {
      return (
        <UserState.Provider value={{ displayName: 'My User' } as any}>
          <BrowserRouter>
            <MenuHeaderInner />
          </BrowserRouter>
        </UserState.Provider>
      );
    }
    const mounted = mount(<WithUser />);
    mounted.setProps({});
    expect(mounted.find('ForwardRef(Avatar)').prop('name')).toBe('My User');
  });

  it('should highlight the tab for the page that is currently active', () => {
    fakeLocation('/');
    const wrapper = shallow(<MenuHeaderInner />);
    const dashbaordLink = wrapper.find('Link').at(0);
    expect(dashbaordLink.prop('className')).toMatchInlineSnapshot(`"item active"`);
  });

  it('should not highlight the tab for pages that are not currently active', () => {
    fakeLocation('/not/the/dashboard');
    const wrapper = shallow(<MenuHeaderInner />);
    const dashbaordLink = wrapper.find('Link').at(0);
    expect(dashbaordLink.prop('className')).toMatchInlineSnapshot(`"item"`);
  });
});

describe('isPathActive', () => {
  it('should be true for exact matches', () => {
    fakeLocation('/abc');
    expect(isPathActive('/abc', true)).toBe(true);
  });

  it('should be false for child paths if exact is enabled', () => {
    fakeLocation('/abc/def');
    expect(isPathActive('/abc', true)).toBe(false);
  });

  it('should be true for exact matches even if exact is disabled', () => {
    fakeLocation('/abc');
    expect(isPathActive('/abc', false)).toBe(true);
  });

  it('should be true for child paths if exact is disabled', () => {
    fakeLocation('/abc/def');
    expect(isPathActive('/abc', false)).toBe(true);
  });
});
