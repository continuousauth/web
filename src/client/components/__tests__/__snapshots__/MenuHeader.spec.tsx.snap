// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuHeader /> Should render a logo 1`] = `
<Memo(ForwardRef(Pane))
  className="container"
>
  <CFALogo
    className="logo"
  />
  <span
    className="productName"
  >
    CFA
  </span>
  <Link
    className="item active"
    to="/"
  >
    Dashboard
  </Link>
  <a
    className="item"
    href="https://docs.continuousauth.dev"
    target="_blank"
  >
    Documentation
  </a>
  <div
    className="item warning"
  >
    <Memo(ForwardRef(WarningSignIcon))
      color="warning"
      marginRight={8}
    />
    CFA is open-source and does not have formal support
  </div>
  <Memo(ForwardRef(Pane))
    className="right"
  >
    <Memo(ForwardRef(Avatar))
      isSolid={true}
      name="?"
      size={24}
    />
    <span>
      ?
    </span>
  </Memo(ForwardRef(Pane))>
  <span />
</Memo(ForwardRef(Pane))>
`;
